libr/include/r_util/r_annotated_code.h

```diff
@@ -0,0 +1,95 @@
+
+#ifndef R2GHIDRA_ANNOTATEDCODE_H
+#define R2GHIDRA_ANNOTATEDCODE_H
+
+#include <r_core.h>
+#include <r_types.h>
+#include <r_vector.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef enum r_syntax_highlight_type_t {
+	R_SYNTAX_HIGHLIGHT_TYPE_KEYWORD,
+	R_SYNTAX_HIGHLIGHT_TYPE_COMMENT,
+	R_SYNTAX_HIGHLIGHT_TYPE_DATATYPE,
+	R_SYNTAX_HIGHLIGHT_TYPE_FUNCTION_NAME,
+	R_SYNTAX_HIGHLIGHT_TYPE_FUNCTION_PARAMETER,
+	R_SYNTAX_HIGHLIGHT_TYPE_LOCAL_VARIABLE,
+	R_SYNTAX_HIGHLIGHT_TYPE_CONSTANT_VARIABLE,
+	R_SYNTAX_HIGHLIGHT_TYPE_GLOBAL_VARIABLE,
+} RSyntaxHighlightType;
+
+typedef enum r_code_annotation_type_t {
+	R_CODE_ANNOTATION_TYPE_OFFSET,
+	R_CODE_ANNOTATION_TYPE_SYNTAX_HIGHLIGHT,
+	// ...
+} RCodeAnnotationType;
+
+typedef struct r_code_annotation_t {
+	size_t start;
+	size_t end;
+	RCodeAnnotationType type;
+	union
+	{
+		struct
+		{
+			ut64 offset;
+		} offset;
+
+		struct
+		{
+			RSyntaxHighlightType type;
+		} syntax_highlight;
+	};
+} RCodeAnnotation;
+
+typedef struct r_annotated_code_t {
+	char *code; // owned
+	RVector/*<RCodeAnnotation>*/ annotations;
+} RAnnotatedCode;
+
+/*
+	Creates a new instance of RAnnotatedCode and returns it.
+*/
+R_API RAnnotatedCode *r_annotated_code_new(char *code);
+/*
+	Deallocate the memory for code
+*/
+R_API void r_annotated_code_free(RAnnotatedCode *code);
+/*
+	Adds annotation to the set of RCodeAnnotations in code
+*/
+R_API void r_annotated_code_add_annotation(RAnnotatedCode *code, RCodeAnnotation *annotation);
+/*
+	Makes an RPVector and inserts the pointers to all annotations that in which annotation->start <= offset < annotation->end
+	This RPVector can have annotations for syntax highlight and for offset.
+*/
+R_API RPVector *r_annotated_code_annotations_in(RAnnotatedCode *code, size_t offset);
+/*
+	Makes an RPVector and inserts pointers to all annotations whose range overlap with range [start, end-1],  and returns it.
+*/
+R_API RPVector *r_annotated_code_annotations_range(RAnnotatedCode *code, size_t start, size_t end);
+/*
+	Takes RAnnotatedCode structure and prints it in JSON format.
+*/
+R_API void r_annotated_code_print_json(RAnnotatedCode *code);
```

Comment:
this should not be here, because this file is for r_util and the code is defined in core. This should be somewhere in in `libr/include`, maybe even in its own files or just directly in `libr/include/r_core.h` but not here.

Reply:
same for the other functions that are defined in `core_annotated_code.c`

Reply:
@thestr4ng3r told me to do it this way. Maybe he overlooked this. I think I should wait for his opinion on this.

Reply:
@ret2libc is right. I would prefix these functions with `r_core_annotated_code_...` and put their declarations right here: https://github.com/radareorg/radare2/blob/fea70914e8b04e10825f6e79c623cdb1b5d49591/libr/include/r_core.h#L922

Reply:
@thestr4ng3r  These printing functions need `RAnnotatedCode` that is is r_annotated_code.h
```
R_API void r_core_annotated_code_print_json(RAnnotatedCode *code);
R_API void r_core_annotated_code_print(RAnnotatedCode *code, RVector *line_offsets);
R_API void r_core_annotated_code_print_comment_cmds(RAnnotatedCode *code);
```
How can we handle this issue?

Reply:
@thestr4ng3r @ret2libc  as I said earlier, the printing function depends on `r_annotated_code.h` and `r_annotated_code.h` depends on `r_core.h`, so we can't move this without changing any of these dependencies. But looking at the code, I don't understand why we need `#include <r_core.h>` in `r_annotated_code.h `. Can we not have that in `r_annotated_code.h `?

Reply:
I am committing with this change now.